rfcs/lower-bounds-calculation.md

```diff
@@ -0,0 +1,215 @@
+
+We need a solution here.
+
+To resolve this, let's reframe things in simpler terms:
+
+If there is never a subtyping relationship between packs of different length, then we don't have any soundness issues, but we find ourselves unable to register `good_callback`.
+
+To resolve _that_, consider that we are in truth being a bit hasty when we say `good_callback : () -> ()`.  We can pass any number of arguments to this function safely.  We could choose to type `good_callback : () -> () & (any) -> () & (any, any) -> () & ...`.  Luau already has syntax for this particular sort of infinite intersection: `good_callback : (any...) -> ()`.
+
+So, we propose some different inference rules for functions:
+
+1. The AST fragment `function(arg0..argN) ... end` is typed `(T0..TN, any...) -> R` where `arg0..argN : T0..TN` and `R` is the inferred return type of the function body.  Function statements are inferred the same way.
+1. Type annotations are unchanged.  `() -> ()` is still a nullary function.
+
+And subtyping rules for type packs:
+
+1. `(T0..TN) <: (U0..UN)` if, for each `T` and `U`, `T <: U`
+
+1. `A & B <: T` if `T <: A` or `T <: B` is the standard left-side subtyping rule for intersections.  It is unchanged.  We include it here for clarity.
+1. `(U...) ~ () & (U) & (U, U) & (U, U, U) & ...`, therefore
+1. `(U...) <: (T0..TN)` if for each `T`, `T <: U`
+
+Notably, we remove all subtyping rules that mention options.  Functions of different arities are no longer ever considered subtypes of one another.  Optional function arguments are now considered solely a feature of function calls.
+
+The standard subtyping rule for functions is unchanged.
+
+1. `T -> R <: U -> S` if `U <: T` and `R <: S`
+
+Under these rules, functions of different arities can never be converted to one another, but actual functions are known to be safe to oversaturate with anything, and so gain a type that says so.
+
+Under these subtyping rules, snippets `C.lua` and `D.lua`, check the way we want: literal functions are implicitly safe to oversaturate, so it is fine to cast them as the necessary callback function type.
+
+`E.lua` also typechecks the way we need it to: `(Instance) -> () </: () -> ()` and so `obscured` cannot receive the value `bad_callback`, which prevents it from being passed to `register_callback`.  However, `good_callback : (any...) -> ()` and `(any...) -> () <: (Instance) -> ()` and so it is safe to register `good_callback`.
+
+Snippet `F.lua` is also fixed with this ruleset: There is no subtyping relationship between `(any) -> ()` and `(any, any) -> ()`, so the intersection is not combined under normalization.
+
+This works, but itself creates some small problems that we need to resolve:
+
+First, the type of `...` still needs to be `()` for functions that have been given this implicit `...any` type. (TODO verify what Lua does in this situation)
```

Comment:
`...` is not in scope outside of variadic functions, so this type pack is not accessible.

Reply:
Looks like we get this from the parser, so there's no problem here.